### ISSUE TITLE
DNS lookups fail inside docker container in AWS govcloud

### ISSUE THREAD

**Issue by chisholm** [sequenceiq/cloudbreak#857]
Cloudbreak isn't working in the AWS "govcloud" (us-gov-west-1) region. The reason is that DNS lookups don't work inside the docker container.  I'm just wondering if anyone has had similar problems or has any insights.  The lookups work fine outside the container.

OS: rhel7
Docker version: 1.7.1
Java: OpenJDK 1.8.0_51

In case it is of interest, the failure occurs when attempting to create a Cloudbreak "credential".  It's trying to connect to the AWS Security Token Service(s) and failing to resolve "sts.amazonaws.com", one of the service endpoints.

Any help appreciated!

**Comment by matyix** [comment]
Hi Chris, 

I'd be surprised if it worked. AFAIK the GovCloud using a different STS endpoint (http://docs.aws.amazon.com/govcloud-us/latest/UserGuide/using-govcloud-endpoints.html).

Janos

**Comment by chisholm** [comment]
Yeah, it's definitely not clear it would work.  But that's a moot point until it can actually resolve the host to an IP and try it.

Fyi, I tracked down [the code](https://github.com/sequenceiq/cloudbreak/blob/master/core/src/main/java/com/sequenceiq/cloudbreak/service/credential/aws/AwsCredentialHandler.java#L84) that triggers this, and it's written to iterate through *all* regions (except govcloud and... north China) and try to import key pairs.  So maybe the import fails N times and doesn't get any keypairs, and it still continues on its merry way and the rest of it works.  I won't know until DNS lookups start working :(

**Comment by matyix** [comment]
DNS lookup definitely should work and uses the cloud providers's DNS to resolve the lookups. If you compare `/etc/consul/consul.json` with `/etc/resolv.conf `you should see the Consul falls back to the same DNS as used by the host. 

```
[cloudbreak@cbgateway-0-azureujteszt-20150812152938 ~]$ cat /etc/consul/consul.json
{
  "bootstrap_expect": 3,
  "server": true,
  "advertise_addr": "10.0.0.5",
  "data_dir": "/data",
  "ui_dir": "/ui",
  "client_addr": "0.0.0.0",
  "recursors": [
    "168.63.129.16"
  ],
  "disable_update_check": true,
  "retry_join": [
    "10.0.0.8",
    "10.0.0.7"
  ],
  "ports": {
    "DNS": 53,
    "HTTP": 8500,
    "HTTPS": -1
  },
  "dns_config": {
    "allow_stale": true,
    "max_stale": "5m",
    "node_ttl": "1m"
  }
}
[cloudbreak@cbgateway-0-azureujteszt-20150812152938 ~]$ cat /etc/resolv.conf
# Generated by NetworkManager
search cbgateway-0-azureujteszt-20150812152938.a7.internal.cloudapp.net
nameserver 168.63.129.16
```

**Comment by chisholm** [comment]
This is what I see in the consul container.  Consul is started as:

    /bin/consul agent -config-dir=/config --server --bootstrap --advertise 172.17.42.1

Its config is:

    bash-4.3# cat /config/consul.json
    {
            "data_dir": "/data",
            "ui_dir": "/ui",
            "client_addr": "0.0.0.0",
            "ports": {
                    "dns": 53
            },
            "recursor": "8.8.8.8",
            "disable_update_check": true
    }

The host's /etc/resolve.conf is:

    $ cat /etc/resolv.conf
    # Generated by NetworkManager
    search us-gov-west-1.compute.internal
    nameserver 172.20.0.2

I don't see 172.20.0.2 in the consul config.  So I don't think they match.  (Maybe hand-editing the "recursor" value would help?  Though I shouldn't need to do that, right?)

**Comment by matyix** [comment]
Looks like you are using an older version of Cloudbreak. 

This was changed a while ago: https://github.com/sequenceiq/munchausen/commit/c3c8ad207f3cf16d440de1ee809c225b1e825e6b 

Can you run `cbd doctor` and check the version, etc? If you upgrade to 1.0.0 (`cbd update`) that should use the correct DNS.

**Comment by chisholm** [comment]
I did a `cbd update` and now have cbd 1.0.0.  `cbd doctor` shows:

    ===> Deployer doctor: Checks your environment, and reports a diagnose.
    local version:1.0.0
    latest release:1.0.0
    docker command: OK
    docker client version: OK
    /tmp/bashenv.259200910: line 1583: [: -lt: unary operator expected
    docker server version: OK
    [WARN] docker-compose.yml already exists, BUT generate would create a DIFFERENT one!
    [WARN] if you want to regenerate, remove it first:
      cbd regenerate
    [WARN] expected change:

(omitted docker-compose.yml change, as it contains secret keys)

The consul docker image it's using is `sequenceiq/consul:v0.5.0-v3`.  The consul config is the same as before and DNS lookups still don't work.  I'm not sure the consul image changed in the upgrade, but I think some other images were updated.  Cloudbreak itself is at 1.0.0.

In case it helps, here are the docker images it's using:

    sequenceiq/uaa:1.8.1-v2
    sequenceiq/socat:latest
    gliderlabs/logspout:master
    sequenceiq/sultans-bin:1.0.1
    sequenceiq/periscope:0.5.6
    sequenceiq/pcdb:0.5.6
    gliderlabs/registrator:v5
    sequenceiq/consul:v0.5.0-v3
    sequenceiq/cbdb:1.0.0
    postgres:9.4.1
    sequenceiq/uluwatu-bin:1.0.1
    sequenceiq/cloudbreak:1.0.0
    sequenceiq/ambassadord:0.5.0

**Comment by matyix** [comment]
The container versions to be used are in the `docker-compose.yml`, thus the file has to be regenerated (as indicated by `cbd doctor` output. Editing the compose file manually it's not a good idea - everything should be added in the Profile and re/generate pick up and adds in the compose file.

```
.....

export AWS_ACCESS_KEY_ID=
export AWS_SECRET_ACCESS_KEY=
```

If you still have problems please join this HipChat room from a browser: https://www.hipchat.com/gvKQLjyea

**Comment by chisholm** [comment]
Yes, I regenerated it and then added the keys.  In fact, I restarted from a blank deployment directory and did `cbd init` and `cbd generate`.  But I do like that you can use the Profile instead so `cbd doctor` doesn't give you dire warnings.  I'll change that.

**Comment by chisholm** [comment]
I eventually figured this out.  I've had to use a VM based on a "hardened" image; the nature of the hardening was not described to me.  But I found some [information](https://docs.docker.com/articles/networking/#communication-between-containers-and-the-wider-world) on the Docker website that gave me what I needed.  The key was the IP forwarding system setting.  It had been switched off; switching it on made it work.

So it was not a cloudbreak problem; I'll close this issue.  Thanks for trying to help!
